### PR TITLE
annotation class

### PR DESCRIPTION
**Pull request by siebrenf**
#154

### CODE REVIEW

File: genomepy/annotation.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+import sys
+from tempfile import mkdtemp
+from typing import Optional, Union
+
+from genomepy.utils import (
+    get_genomes_dir,
+    gzip_and_name,
+    gunzip_and_name,
+    glob_ext_files,
+    update_readme,
```
Comment (siebrenf) [inline genomepy/annotation.py:15]: can ignore `update_readme` for code review

File: genomepy/annotation.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+import subprocess as sp
+import sys
+from tempfile import mkdtemp
+from typing import Optional, Union
+
+from genomepy.utils import (
+    get_genomes_dir,
+    gzip_and_name,
+    gunzip_and_name,
+    glob_ext_files,
```
Comment (siebrenf) [inline genomepy/annotation.py:14]: Return (gzipped) file names in directory containing the given extension.

File: genomepy/annotation.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+import re
+import shutil
+import subprocess as sp
+import sys
+from tempfile import mkdtemp
+from typing import Optional, Union
+
+from genomepy.utils import (
+    get_genomes_dir,
+    gzip_and_name,
```
Comment (siebrenf) [inline genomepy/annotation.py:12]: Gzip file (argument 1) if requested (argument 2) & returns up to date filename

File: genomepy/annotation.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    gunzip_and_name,
```
Comment (siebrenf) [inline genomepy/annotation.py:13]: Unzip file (argument 1) & returns up to date filename and if it was unzipped

File: genomepy/annotation.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        If the contig names do conform (after fixing), filter the annotation contigs for genome contigs.
+        Log the results in the README.
+
+        filter_contigs: remove contigs from the annotations that are missing from the genome?
+        """
+        if self.genome_file is None:
+            sys.stderr.write("A genome is required for sanitizing!")
+            return
+
+        if self._is_conforming():
```
Comment (simonvh) [inline genomepy/annotation.py:293]: To me the `conforming` terminology is not immediately clear, what is meant by that?

File: genomepy/annotation.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+            )
+        if check_exists:
+            raise FileNotFoundError(
+                f"could not find '{self.name}.{ext}(.gz)' in directory {self.genome_dir}"
+            )
+        return None
+
+    @property
+    def genome_contigs(self):
+        return self.__genome_contigs
```
Comment (Maarten-vd-Sande) [inline genomepy/annotation.py:77]: any reason to use double underscore (`_`) here? I don't know what is more correct, but I believe accessing the double underscored attrs is much more difficult outside of the class than the single one

File: genomepy/annotation.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        # update documentation
+        status = "sanitized"
+        extra_lines = []
+        if missing_contigs and filter_contigs:
+            status = "sanitized and filtered"
+            extra_lines = [
+                "",
+                "The following contigs were filtered out of the gene annotation:",
+                f"{', '.join(missing_contigs)}",
+            ]
```
Comment (simonvh) [inline genomepy/annotation.py:345]: This seems to be duplicated with lines 302-306

File: genomepy/annotation.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    """
+    if mode not in ["r", "w"]:
+        raise ValueError("mode must be either 'r' or 'w'.")
+
+    if fname.endswith(".gz"):
+        return gzip.open(fname, mode + "t")
+    return open(fname, mode)
+
+
+def get_column(
```
Comment (simonvh) [inline genomepy/annotation.py:371]: pandas? or don't you want an (extra, maybe unnecessary) dependency?

File: genomepy/annotation.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+        return list(set(missing_contigs))
+
+    def _is_conforming(self):
+        """
+        Check if genome and annotation contigs conform.
+        Returns bool.
+        """
+        genome_contigs = self.genome_contigs
+        annotation_contigs = self.annotation_contigs
+        conforming = bool(set(genome_contigs) & set(annotation_contigs))
```
Comment (Maarten-vd-Sande) [inline genomepy/annotation.py:167]: You can use the self.attrs directly here

Comment (Maarten-vd-Sande) [reply genomepy/annotation.py:167]:
```
def _is_conforming(self):
    return bool(set(self.genome_contigs) & set(self.annotation_contigs))
```

Comment (Maarten-vd-Sande) [reply genomepy/annotation.py:167]: What does it even do? Check if there are any overlapping contigs between the genome and annotaion?

Comment (siebrenf) [reply genomepy/annotation.py:167]: yes!

File: genomepy/annotation.py [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        os.replace(new_gtf_file, self.annotation_gtf_file)
+        rm_rf(tmp_dir)
+        return list(set(missing_contigs))
+
+    def bed_from_gtf(self):
+        """
+        Create an annotation bed file from an annotation gtf file.
+        Overwrites the existing bed file.
+        """
+        tmp_dir = mkdtemp(dir=self.genome_dir)
```
Comment (simonvh) [inline genomepy/annotation.py:258]: do you want to do this the genome_dir? if there is an error, or the program is interrupted, there will be a temp dir there. Why not use the system tempdir?

File: genomepy/annotation.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+        rm_rf(tmp_dir)
+        return list(set(missing_contigs))
+
+    def filter_genome_contigs(self):
+        """
+        filter the gtf for contigs in the genome.
+        return a list of discarded contigs.
+        """
+        missing_contigs = []
+        genome_contigs = self.genome_contigs
```
Comment (Maarten-vd-Sande) [inline genomepy/annotation.py:140]: I think this is kinda "dangerous". `genome_contigs` is a list, which means it is mutable, which in turn means that the = does not mean it takes a copy. Both self.genome_contigs and genome_contigs are exactly the same variable, just with a different name. If you now start changing genome_contigs, it means you change self.genome_contigs.

File: genomepy/annotation.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+        self.annotation_gtf_file = gzip_and_name(self.annotation_gtf_file, is_unzipped)
+        new_bed_file = gzip_and_name(
+            new_bed_file, self.annotation_bed_file.endswith(".gz")
+        )
+
+        os.replace(new_bed_file, self.annotation_bed_file)
+        rm_rf(tmp_dir)
+
+    def sanitize(self, filter_contigs: Optional[bool] = True):
+        """
```
Comment (simonvh) [inline genomepy/annotation.py:281]: Function can be simplified by splitting it out more, for me it is hard to follow now. It should be three steps, correct?
1) sanitize if necessary
2) remove annotation not in reference genome
3) update README

However, this logic is mixed now, with three returns in the function
